### ISSUE TITLE
Apply localization to all hard coded text in the Expensify.cash app

### ISSUE THREAD

**Issue by michaelhaxhiu** [Expensify/Expensify.cash#2408]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/master/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!

Upwork job - https://www.upwork.com/jobs/~01ad599888651ad174
___

## Expected Result:
We are looking to apply localization to the Expensify.cash app, so that all hard coded text is replaced with locale-aware components/ methods. This project will require (1) Identifying all places that display hardcoded texts in the app, and (2) Using the `withLocalize` HOC in this PR https://github.com/Expensify/Expensify.cash/pull/2208 to make all those texts support localized versions.

## Actual Result:
Today, all text in the Expensify.cash app is hardcoded in English. Here are a few examples for reference:

<img width=300 src=https://user-images.githubusercontent.com/13206896/114899231-c6e40700-9de0-11eb-8d92-6670b3d105d6.png>

![image](https://user-images.githubusercontent.com/13206896/114883717-2c30fb80-9dd3-11eb-8ff8-467b808bd62d.png)

![image](https://user-images.githubusercontent.com/13206896/114883729-2e935580-9dd3-11eb-806d-5b56a294a70c.png)

## Platform:
This should apply to all platforms listed:

- [ ] Web
- [ ] iOS
- [ ] Android
- [ ] Desktop App
- [ ] Mobile Web

**Expensify/Expensify Issue URL:** https://github.com/Expensify/Expensify/issues/152153

## Other considerations:
There may be a few areas where it doesn't make sense to translate hardcoded text, which are important to highlight as a part of this project. For example, we should not update the hard corded text found here:

- Click avatar (above chat list) > Profile > Set my time zone manually > see dropdown list of regions
![exemption](https://user-images.githubusercontent.com/13206896/114900836-3c040c00-9de2-11eb-96fa-5e8fc2f838a1.gif)

**Comment by michaelhaxhiu** [comment]
Posting to Upwork right now - link incoming.

**Comment by michaelhaxhiu** [comment]
Posted to Upwork (job link in GH body) and awaiting proposals.

**Comment by eVoloshchak** [comment]
### Proposal 

1. In every component we will use withLocalize HOC from components/withLocalize.js;
2. I can see that all localization library setup is already done, there are two localization files in "languages" directory;
3. Since there is no Spanish translation, the es.js locale file will contain placeholder english localization from en.js (to avoid "missing translation" errors);
4. In each component translation function will be used as such:
const { translations: { translate } } = withLocalize();
And all of the instances of <Text>Phrase</Text> will be substituted with <Text>{translate('phrase)}</Text>

### Question
Am I correct in thinking that all of the files that need localization are contained in "components" and "pages" directories?
Or is there such files in other directories too?

**Comment by eVoloshchak** [comment]
> Don't do this, we are not translating the app to spanish yet (or any other language) but also we don't want to leave garbage in the translation files.

Got it, thank you for clarification

> Huh? Why? That's not how you use the higher order component, its usage is similar to the withOnyx higher order component.

My bad, I confused HOC with hook. I remember being slightly confused seeing withLocalize in "components" directory, thinking it was a hook. I got it now. I also see the "compose" function from redux.

**Comment by iwiznia** [comment]
Can you list out the places we will update? After reviewing that I will approve your proposal.

**Comment by eVoloshchak** [comment]
> Can you list out the places we will update? After reviewing that I will approve your proposal.

Here is the list of components that need to be updated:
Components

    AttachmentPicker (native)
    TextInputFocusable
    BaseUpdateAppModal
    AttachmentModal
    ConfirmModal
    HeaderWithCloseButton
    IOUConfirmationList
    OptionsSelector
    RenderHTML
    UnreadActionIndicator
    VideoChatButtonAndMenu
    WelcomeText

Pages
    home
        EmojiPickerMenu
        ReportActionCompose
        ReportActionContextMenu
        ReportActionItemFragment
        ReportActionsView
        ReportTypingIndicator
        SidebarScreen
    iou
        IOUParticipantsRequest
        IOUParticipantsSplit
        IOUAmountPage
        IOUModal
    setings
        LoginField
        ProfilePage
        AddSecondaryLoginPage
        InitialSettingsPage
        PasswordPage
        PaymentsPage
        PreferencesPage
    signin
        LoginFormNarrow
        LoginFormWide
        SignInPageLayoutNarrow
        SignInPageLayoutWide
        TermsOnly
        TermsWithLicenses
        ChangeExpensifyLoginLink
        PasswordForm
        ResendValidationForm

    DetailsPage
    NewChatPage
    NewGroupPage
    NotFound
    ReportParticipantsPage
    SearchPage
    SetPasswordPage
    
    
I have another question: what should we do with the text labels in defaultProps or in variables set outside of the react components? These won't be able to use 'translate' function. Should i move all of the default props and constants inside the corresponding react componets?

**Comment by iwiznia** [comment]
Some of those don't seem to have any hardcoded text:
> TextInputFocusable
> HeaderWithCloseButton
> EmojiPickerMenu
> ReportActionItemFragment
> ReportActionsView

I don't see `InitialSettingsPage` did you mean `InitialPage`?

> I have another question: what should we do with the text labels in defaultProps or in variables set outside of the react components? These won't be able to use 'translate' function. Should i move all of the default props and constants inside the corresponding react componets?

Great question! I think what you need to do there is basically move the translation to the render method. So, for example [here](https://github.com/Expensify/Expensify.cash/blob/75c554970f89b926e45f26b15acd2ea10da93ef3/src/components/OptionsSelector.js#L66) you would:
1. Make the default prop value an empty string
2. In the render method, where we use it, do something like: `this.props.placeholderText || translate(keyHere)` 

That way the translation is moved from initialization to the render method and it can get updated whenever the locale changes. 
Does that make sense?

**Comment by eVoloshchak** [comment]
> Some of those don't seem to have any hardcoded text:
TextInputFocusable - lines 182, 190
HeaderWithCloseButton - line 66
EmojiPickerMenu - line 104 
ReportActionItemFragment - lines 68-82
ReportActionsView- line 303

> I don't see InitialSettingsPage did you mean InitialPage?
Yes, exactly. Component with dislay name 'InitialSettingsPage' inside the src/pages/settings/InitialPage.js

> That way the translation is moved from initialization to the render method and it can get updated whenever the locale changes. Does that make sense?
Yes, it makes perfect sense. Thank you for clarification

**Comment by iwiznia** [comment]
Oh, sorry about that. You are right about all those having texts, I missed it. The only one I am having doubts is [this one](https://github.com/Expensify/Expensify.cash/blob/a75a4077b526417c09ed07316334a707cd2feea2/src/pages/home/report/ReportActionItemFragment.js#L68), is that actually printed in screen for the user to see?

> There is also variables defined outside the component`s scope, for instance CONTEXT_ACTIONS in ReportActionContextMenu and messages and pronouns in CONST.js. How should we handle these?

Yeah, for those you will need to subscribe to the PREFERRED_LOCALE onyx key and call the [translate](https://github.com/Expensify/Expensify.cash/blob/b3172945b0ff4d9a0a223ec574706d1ed9601f40/src/libs/translate.js#L15) method directly.

@michaelhaxhiu go ahead and hire @eVoloshchak in upwork.

**Comment by parasharrajat** [comment]
@eVoloshchak 
> There are also variables defined outside the component`s scope, for instance, CONTEXT_ACTIONS in ReportActionContextMenu and messages and pronouns in CONST.js. How should we handle these?

These should be used somewhere in any component and I think you can call the `translate` from the HOC there. can't we?

**Comment by michaelhaxhiu** [comment]
@eVoloshchak Hired you on Upwork for the job, and assigned you to this GH. ✅  Feel free to begin drafting a PR for the proposed solution. 

Given that this is your first time working on an Expensify.cash job, I wanted to highlight the [Contributor Guidelines]
(https://github.com/Expensify/Expensify.cash/blob/master/CONTRIBUTING.md) in case you haven't read it yet!

**Comment by iwiznia** [comment]
> These should be used somewhere in any component and I think you can call the translate from the HOC there. can't we?

I think that won't work since we will have the text, not the key.

**Comment by parasharrajat** [comment]
so can't we set the key as CONST values and thus they would be passed down.
Like
```
const CONTEXT_ACTIONS = [
    // Copy to clipboard
    {
        text: 'Copy to Clipboard', ==> Key here
        icon: ClipboardIcon,
        successText: 'Copied!',
        successIcon: Checkmark,
        shouldShow: true,
```

**Comment by iwiznia** [comment]
True, maybe that's better.

**Comment by eVoloshchak** [comment]
> The only one I am having doubts is this one, is that actually printed in screen for the user to see?

Judging by the source code it is visible to user on the screen, but I will have to double check when I actually launch the application.

> Feel free to begin drafting a PR for the proposed solution. Given that this is your first time working on an Expensify.cash job, I wanted to highlight the [Contributor Guidelines]
(https://github.com/Expensify/Expensify.cash/blob/master/CONTRIBUTING.md) in case you haven't read it yet!

Starting to work on the PR, thank you for the guidelines

> so can't we set the key as CONST values and thus they would be passed down.

Got it

**Comment by michaelhaxhiu** [comment]
@eVoloshchak how's the PR coming along so far? Feel free to provide daily updates as you go so that we can have a gauge of your progress.

**Comment by eVoloshchak** [comment]
> @eVoloshchak how's the PR coming along so far? Feel free to provide daily updates as you go so that we can have a gauge of your progress.

I have applied localization to a majority of components/pages, I have 4 of them left to be translated. I expect to finish them in a couple of hours, after that i will start testing on different platforms/devices. If tests are successfull, I'll submit a PR today.

**Comment by iwiznia** [comment]
I reverted the PR (left a comment in the PR itself)

**Comment by eVoloshchak** [comment]
> I reverted the PR (left a comment in the PR itself)

I've resolved some issues with the PR, but I still have a couple of questions. I've asked them here: https://github.com/Expensify/Expensify.cash/pull/2562

**Comment by michaelhaxhiu** [comment]
What is the latest on this one?

**Comment by iwiznia** [comment]
@michaelhaxhiu we reverted the PR ad @eVoloshchak was waiting on some answers from me which they did not get till today since I was OOO.

**Comment by iwiznia** [comment]
@eVoloshchak can you confirm you are working on this?

**Comment by eVoloshchak** [comment]
> @eVoloshchak can you confirm you are working on this?

Yes, I'm working on this, planning to submit a PR in a couple of hours.

**Comment by Julesssss** [comment]
Hi @eVoloshchak, I think this PR has caused a regression with the IOU Modal, but this wasn't possible for you to test, so no worries 🙂

I'm seeing a crash [here](https://github.com/Expensify/Expensify.cash/blob/cd915cd9083821bc8f3576288e35246075f4def5/src/pages/iou/steps/IOUAmountPage.js#L60) which doesn't occur on previous commits:
```
Uncaught TypeError: this.textInput.focus is not a function
    at IOUAmountPage.componentDidMount (IOUAmountPage.js:60)
    at commitLifeCycles (react-dom.development.js:19814)
    at commitLayoutEffects (react-dom.development.js:22803)
    at HTMLUnknownElement.callCallback (react-dom.development.js:188)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:237)
    at invokeGuardedCallback (react-dom.development.js:292)
    at commitRootImpl (react-dom.development.js:22541)
    at unstable_runWithPriority (scheduler.development.js:653)
    at runWithPriority$1 (react-dom.development.js:11039)
    at commitRoot (react-dom.development.js:22381)
```

As the feature is currently behind a beta, you must perform the following change to reproduce the issue. Either navigate to Permissions.js,[ comment out this line ](https://github.com/Expensify/Expensify.cash/blob/e28b50e139c492442ea984c3d48e23be744ceb4e/src/libs/Permissions.js#L31)and return true -- or apply this [diff.txt](https://github.com/Expensify/Expensify/files/6427621/diff.txt) with ` git apply path/to/diff.txt`

**Comment by eVoloshchak** [comment]
> There's also [this other error](https://expensify.slack.com/archives/C01GTK53T8Q/p1620683353027200): `Error: addSecondaryLoginPage.password was not found in the default language` which we need to fix too.
> @eVoloshchak are you on it?

I have fixed the error with addSecondaryLoginPage.password, ready to submit a PR.
Do I just create a new PR referencing this issue?
I can see that there is already a [pull request](https://github.com/Expensify/Expensify.cash/pull/2800) that resolves [this](https://github.com/Expensify/Expensify.cash/issues/2786) issue. Should I carry on with fixing it?

**Comment by iwiznia** [comment]
You can't add a reviewer? Odd.. anyway, I see the PR above, so reviewing it now.

**Comment by iwiznia** [comment]
Reopening since a couple of issues were found

**Comment by michaelhaxhiu** [comment]
How long ago were the issues found @iwiznia? This contributor has already been compensated for the job and so I'm thinking it may be better to open a fresh GH.

**Comment by iwiznia** [comment]
I know that we already paid, but the issues were there since the beginning and I expect @eVoloshchak to fix them.

**Comment by michaelhaxhiu** [comment]
Ah, I understand! Ok let's keep this GH going in that case 👍
